gds/src/main/scala/edu/gemini/aspen/gds/observations/ObservationFSM.scala

```diff
@@ -0,0 +1,115 @@
+package edu.gemini.aspen.gds.observations
+
+import cats.effect._
+import cats.effect.std.QueueSink
+import cats.effect.syntax.all._
+import cats.syntax.all._
+import edu.gemini.aspen.gds.syntax.all._
+import edu.gemini.aspen.giapi.data.{ DataLabel, ObservationEvent }
+import java.util.logging.Logger
+import scala.concurrent.duration._
+
+trait ObservationFSM[F[_]] {
+  def addObservationEvent(obsEvent: ObservationEvent): F[Unit]
+  def stopObservation: F[Unit]
+  def step: F[Unit]
+}
+
+object ObservationFSM {
+  private val logger = Logger.getLogger(this.getClass.getName)
+
+  // TODO: make these configurable
+  val numAttempts = 3
+  val sleepTime   = 5.seconds
+
+  def apply[F[_]: Async](
+    dataLabel: DataLabel,
+    obsStateQ: QueueSink[F, ObservationStateEvent]
+  ): F[ObservationFSM[F]] =
+    Ref.of[F, State](Running(Set.empty)).map { state =>
+      new ObservationFSM[F] {
+        val allEvents: Set[ObservationEvent] = ObservationEvent.values.toSet
+
+        def addObservationEvent(obsEvent: ObservationEvent): F[Unit] = state.modify {
+          case st @ Running(events) =>
+            if (events.contains(obsEvent))
+              st -> logger.warningF(
+                s"Duplicate event $obsEvent received for running observation $dataLabel"
+              )
+            else if (obsEvent == ObservationEvent.OBS_END_DSET_WRITE)
```

Comment:
Should we use typesafe equals `===`

Reply:
Yes, I should add `Eq` instances and use `===` for all my comparisons. I just got lazy. I'll fix that.